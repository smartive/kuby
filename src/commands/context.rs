use colored::Colorize;
use dialoguer::Select;
use dialoguer::theme::ColorfulTheme;

use crate::kubeconfig::{load, save};

pub async fn context(name: &Option<String>) -> Result<(), Box<dyn std::error::Error>> {
    let mut config = load()?;
    let contexts: Vec<String> = config.contexts.iter().map(|c| c.name.to_string()).collect();
    let current_context = config.current_context.unwrap_or_else(|| "".to_string());

    let new_context = match name {
        Some(name) => name.to_string(),
        None => {
            let selected = Select::with_theme(&ColorfulTheme::default())
                .with_prompt(format!("Select new context (current: {})", current_context.green()))
                .default(0)
                .items(&contexts)
                .interact()?;
            contexts.get(selected).unwrap_or(&"".to_string()).to_string()
        }
    };

    if current_context == *new_context {
        println!("{} is already set as context.", current_context.green());
        return Ok(());
    }

    println!("Set {} as active context.", new_context.green());
    config.current_context = Some(new_context);
    save(&config)?;

    Ok(())
}
