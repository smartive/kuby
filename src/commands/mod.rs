use clap::Subcommand;

pub mod context;

const CONTEXT_LONG_ABOUT: &str = r#"This command lets you switch your current Kubernetes context.
When you provide a new context (via the name attribute) the new context
is directly selected if possible. If you omit the name,
a list of avilabile contexts is shown and you can select one."#;

#[derive(Subcommand)]
pub enum Commands {
    #[clap(about = "Directly switch a context or show a list of available context", long_about = CONTEXT_LONG_ABOUT)]
    #[clap(visible_alias = "ctx", visible_alias = "c")]
    Context {
        /// The new context to switch to.
        name: Option<String>
    },
}
