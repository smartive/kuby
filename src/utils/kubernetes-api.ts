import { Core_v1Api, KubeConfig } from '@kubernetes/client-node';
import { Context } from '@kubernetes/client-node/dist/config_types';

export class KubernetesApi {
  public static contextOverride: string | undefined;

  public static namespaceOverride: string | undefined;

  /**
   * Manage elements of the kubernetes core api.
   */
  public readonly core: Core_v1Api;

  /**
   * Manage kube-config.
   */
  public readonly kubeConfig: KubeConfig;

  /**
   * Convenience getter for current context.
   *
   * @throws {Error} When the returned context from the config is `null`.
   */
  public get currentContext(): Context {
    const contextName = this.kubeConfig.getCurrentContext();
    const context = this.kubeConfig.getContextObject(contextName);
    if (context == null) {
      throw new Error(`No context with name '${contextName}' found.`);
    }
    return context;
  }

  /**
   * Convenience getter for current namespace. Does take overrides into account (--ns).
   * Returns "default" when no "current" namespace is set
   * on the context.
   */
  public get currentNamespace(): string {
    return KubernetesApi.namespaceOverride || this.currentContext.namespace || 'default';
  }

  private constructor(kubeConfig: KubeConfig) {
    this.kubeConfig = kubeConfig;
    if (KubernetesApi.contextOverride) {
      this.kubeConfig.setCurrentContext(KubernetesApi.contextOverride);
    }

    this.core = this.kubeConfig.makeApiClient(Core_v1Api);
  }

  /**
   * Creates a kubernetes api from the default configuration
   * (located at `~/.kube/config` or the env var that points
   * to the kube config).
   */
  public static fromDefault(): KubernetesApi {
    const kubeConfig = new KubeConfig();
    kubeConfig.loadFromDefault();
    return new KubernetesApi(kubeConfig);
  }

  /**
   * Creates the kubernetes api from a given string content.
   */
  public static fromString(content: string): KubernetesApi {
    const kubeConfig = new KubeConfig();
    kubeConfig.loadFromString(content);
    return new KubernetesApi(kubeConfig);
  }
}
