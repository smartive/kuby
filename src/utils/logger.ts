import * as chalk from 'chalk';

export enum LogLevel {
  debug,
  info,
  warn,
  error,
  output,
}

const ora = require('ora');

export class Logger {
  public static level: LogLevel = LogLevel.info;
  private spinner: typeof ora | undefined;

  constructor(private prefix: string = '') {}

  public startSpinner(
    initialText: string,
    level: LogLevel = LogLevel.info,
  ): void {
    if (level < Logger.level) {
      return;
    }
    this.spinner = ora(initialText).start();
  }

  public stopSpinner(): void {
    if (!this.spinner) {
      return;
    }
    this.spinner.stop();
    delete this.spinner;
  }

  public setSpinnerText(text: string): void {
    if (!this.spinner) {
      return;
    }
    setTimeout(() => {
      if (this.spinner) {
        this.spinner.text = text;
      }
    },         0);
  }

  public spinnerSuccess(text?: string): void {
    if (!this.spinner) {
      return;
    }
    this.spinner.succeed(text);
    delete this.spinner;
  }

  public spinnerFail(text?: string): void {
    if (!this.spinner) {
      return;
    }
    this.spinner.fail(text);
    delete this.spinner;
  }

  public log(level: LogLevel, message: string): void {
    if (level < Logger.level || this.spinner) {
      return;
    }
    this.getLogFunction(level)(`${this.getLevelPrefix(level)}${message}`);
  }

  public debug(message: string): void {
    this.log(LogLevel.debug, message);
  }

  public info(message: string): void {
    this.log(LogLevel.info, message);
  }

  public output(message: string): void {
    this.log(LogLevel.output, message);
  }

  public success(message: string): void {
    this.log(LogLevel.info, chalk.green(message));
  }

  public warn(message: string): void {
    this.log(LogLevel.warn, chalk.yellow(message));
  }

  public error(message: string): void {
    this.log(LogLevel.error, chalk.red(message));
  }

  private getLevelPrefix(level: LogLevel): string {
    if (!this.prefix) {
      return '';
    }
    switch (level) {
      case LogLevel.debug:
        return chalk.grey(`[${this.prefix}]: `);
      case LogLevel.warn:
        return chalk.yellow(`[${this.prefix}]: `);
      case LogLevel.error:
        return chalk.red(`[${this.prefix}]: `);
      default:
        return chalk.blue(`[${this.prefix}]: `);
    }
  }

  private getLogFunction(level: LogLevel): (message: string) => void {
    switch (level) {
      case LogLevel.debug:
        return console.debug;
      case LogLevel.warn:
        return console.warn;
      case LogLevel.error:
        return console.error;
      default:
        return console.info;
    }
  }
}
