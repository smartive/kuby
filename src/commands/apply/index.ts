import { pathExists, readdir, stat } from 'fs-extra';
import { Arguments, Argv, CommandModule } from 'yargs';

import { RootArguments } from '../../root-arguments';
import { ExitCode } from '../../utils/exit-code';
import { Logger } from '../../utils/logger';
import { RcFile } from '../../utils/rc-file';
import { spawn } from '../../utils/spawn';
import { kubeConfigCommand } from '../kube-config';

type ApplyArguments = RootArguments & { deployFolder: string };

export const applyCommand: CommandModule<RootArguments, ApplyArguments> = {
  command: 'apply [deployFolder]',
  describe: 'Apply all prepared yaml files with kubectl.',

  builder: (argv: Argv<RootArguments>) =>
    argv
      .positional('deployFolder', {
        default: './deployment/',
        description: 'Folder with prepared yaml (k8s) files.',
        type: 'string',
        normalize: true,
      })
      .completion('completion', undefined, async (_, argv: Arguments) => {
        if (argv._.length >= 3) {
          return [];
        }
        const dirs = [];
        const directory = await readdir(process.cwd());
        for (const path of directory) {
          const stats = await stat(path);
          if (stats.isDirectory()) {
            dirs.push(path);
          }
        }
        return dirs;
      }),

  async handler(args: Arguments<ApplyArguments>): Promise<void> {
    if (args.getYargsCompletions) {
      return;
    }

    const logger = new Logger('deployment');
    logger.debug('Apply yaml files');

    if (!(await pathExists(args.deployFolder))) {
      logger.error('Deploy directory does not exist. Aborting.');
      process.exit(ExitCode.error);
      return;
    }

    if (args.ci) {
      await kubeConfigCommand.handler({
        ...args,
        noInteraction: true,
        force: true,
      });
    }

    const code = await spawn('kubectl', RcFile.getKubectlArguments(args, ['apply', '-f', args.deployFolder]));

    if (code !== 0) {
      logger.error('An error happend during the kubectl command.');
      process.exit(ExitCode.error);
      return;
    }

    logger.success('Files applied.');
  },
};
