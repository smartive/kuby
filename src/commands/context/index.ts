import * as chalk from 'chalk';
import { prompt } from 'inquirer';
import { Arguments, Argv, CommandModule } from 'yargs';

import { RootArguments } from '../../root-arguments';
import { exec } from '../../utils/exec';
import { ExitCode } from '../../utils/exit-code';
import { Logger } from '../../utils/logger';
import { getContexts, getCurrentContext } from './utils/kubectx';

const fuzzy = require('fuzzy');

export const contextCommand: CommandModule<
  RootArguments,
  {
    name?: string;
  }
> = {
  command: 'context [name]',
  aliases: 'ctx',
  describe:
    'Utilities for kubernetes contexts. Without any subcommands, ' + 'prints a list of possible configured contexts.',

  builder: (argv: Argv) =>
    argv
      .positional('name', {
        description: 'Context to switch to. If omitted, user is asked.',
        type: 'string',
      })
      .completion('completion', false as any, async (_, argv: Arguments) => (argv._.length >= 3 ? [] : await getContexts())),

  async handler(
    args: Arguments<{
      name?: string;
    }>,
  ): Promise<void> {
    if (args.getYargsCompletions) {
      return;
    }

    const logger = new Logger('contexts');
    logger.debug('list / switch contexts');

    const current = await getCurrentContext();
    const contexts = await getContexts();

    if (!args.name) {
      args.name = ((await prompt([
        {
          type: 'autocomplete',
          name: 'context',
          message: `Which context do you want to use? ${chalk.dim(`(current: ${current})`)}`,
          source: async (_: any, input: string) => {
            if (!input) {
              return contexts;
            }
            return fuzzy
              .filter(input.replace(/\s/g, ''), contexts, {
                pre: '{',
                post: '}',
              })
              .map((e: any) => ({
                name: e.string.replace(/{(.)}/g, chalk.underline('$1')),
                value: e.original,
              }));
          },
        } as any,
      ])) as { context: string }).context;
    }

    if (current === args.name) {
      logger.info('No different context selected, exiting.');
      return;
    }

    if (!contexts.includes(args.name)) {
      logger.error(`The context "${args.name}" does not exist.`);
      process.exit(ExitCode.error);
      return;
    }

    await exec(`kubectl config use-context "${args.name}"`);
    logger.info(`Active context is now "${chalk.green(args.name)}".`);
  },
};
