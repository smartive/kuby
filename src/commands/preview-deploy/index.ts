import { V1Namespace, V1ObjectMeta } from '@kubernetes/client-node';
import { async } from 'fast-glob';
import { readFile } from 'fs-extra';
import { EOL } from 'os';
import { posix } from 'path';
import { Arguments, Argv, CommandModule } from 'yargs';

import { RootArguments } from '../../root-arguments';
import { envsubst } from '../../utils/envsubst';
import { exec } from '../../utils/exec';
import { ExitCode } from '../../utils/exit-code';
import { KubernetesApi } from '../../utils/kubernetes-api';
import { Logger } from '../../utils/logger';
import { kubeConfigCommand } from '../kube-config';

type PreviewDeployArguments = RootArguments & {
  name: string;
  sourceFolder: string;
  prefix: string;
  kubeConfig?: string;
};

const logger = new Logger('preview deployment');

async function createNamespace(api: KubernetesApi, name: string): Promise<void> {
  try {
    logger.info(`Create preview namespace "${name}".`);
    await api.core.readNamespace(name);
    logger.info(`Namespace "${name}" already exists.`);
  } catch (e) {
    if (e.response.statusCode !== 404) {
      throw e;
    }

    await api.core.createNamespace({
      ...new V1Namespace(),
      apiVersion: 'v1',
      kind: 'Namespace',
      metadata: {
        ...new V1ObjectMeta(),
        name,
      },
    });
    logger.success(`Created namespace "${name}".`);
  }
}

export const previewDeployCommand: CommandModule<RootArguments, PreviewDeployArguments> = {
  command: 'preview-deploy <name> [sourceFolder]',
  aliases: 'prev-dep',
  describe:
    'Create opinionated preview deployment. This command creates a namespace ' +
    'and deploys the application in the given namespace.',

  builder: (argv: Argv<RootArguments>) =>
    argv
      .option('prefix', {
        string: true,
        description: 'Defines the prefix for the namespace name.',
        default: 'prev-',
      })
      .option('kube-config', {
        string: true,
        description: 'Define the kube config to use for namespace creation. Defaults to $KUBE_CONFIG variable.',
      })
      .positional('name', {
        description: 'Name of the preview namespace.',
        type: 'string',
      })
      .positional('sourceFolder', {
        description: 'Folder to search for yaml files.',
        type: 'string',
        default: './k8s/',
      }) as Argv<PreviewDeployArguments>,

  async handler(args: Arguments<PreviewDeployArguments>): Promise<void> {
    if (args.getYargsCompletions) {
      return;
    }

    logger.debug('Execute preview deployment');

    const name = `${args.prefix}${args.name}`;
    if (name.length > 64) {
      logger.error(`Name of the namespace "${name}" is too long (max: 64 characters, is: ${args.name.length})`);
      process.exit(ExitCode.error);
      return;
    }

    const api = args.kubeConfig ?
      KubernetesApi.fromString(args.kubeConfig.isBase64() ? args.kubeConfig.base64Decode() : args.kubeConfig) :
      KubernetesApi.fromDefault();

    await createNamespace(api, name);

    const files = await async<string>(['**/*.{yml,yaml}'], {
      cwd: args.sourceFolder,
    });
    logger.debug(`Found ${files.length} files for processing.`);

    const yamls = (await Promise.all(files.map(async f => await readFile(posix.join(args.sourceFolder, f), 'utf8'))))
      .map(yaml => envsubst(yaml))
      .join(`${EOL}---${EOL}`);

    await kubeConfigCommand.handler({
      ...args,
      configContent: args.kubeConfig,
      noInteraction: true,
      force: args.ci,
    });

    logger.debug('Executing <echo "templates" | kubectl -n <name> apply -f ->');
    try {
      const result = await exec(`echo "${yamls}" | kubectl -n ${name} apply -f -`);
      logger.info(result);
    } catch (e) {
      logger.error(`Error: ${e}`);
      process.exit(ExitCode.error);
      return;
    }

    logger.success(`Preview Deployment applied to namespace "${name}".`);
  },
};
