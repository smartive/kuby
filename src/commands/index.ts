import { CommandModule } from 'yargs';

import { applyCommand } from './apply';
import { base64Command } from './base64';
import { contextCommand } from './context';
import { deleteCommand } from './delete';
import { deployCommand } from './deploy';
import { kubeConfigCommand } from './kube-config';
import { kubectlCommand } from './kubectl';
import { namespaceCommand } from './namespace';
import { prepareCommand } from './prepare';
import { previewDeployCommand } from './preview-deploy';
import { secretCommand } from './secret';
import { versionCommand } from './version';

export const commands: CommandModule<any, any>[] = [
  applyCommand,
  base64Command,
  contextCommand,
  deleteCommand,
  deployCommand,
  kubeConfigCommand,
  kubectlCommand,
  namespaceCommand,
  prepareCommand,
  previewDeployCommand,
  secretCommand,
  versionCommand,
];
