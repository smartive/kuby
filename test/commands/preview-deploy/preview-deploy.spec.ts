import { vol } from 'memfs';

import { previewDeployCommand } from '../../../src/commands/preview-deploy';
import { exec } from '../../../src/utils/exec';
import { KubernetesApi } from '../../../src/utils/kubernetes-api';
import { Logger } from '../../../src/utils/logger';
import { clearGlobalMocks } from '../../helpers';

function defaultFiles(): void {
  vol.fromJSON({
    '/foo.yml': 'foo:\n  bar: yaml',
    '/bar.yml': 'bar:\n  foo: yaml',
  });
}

describe('commands / preview-deploy', () => {
  beforeAll(() => {
    process.exit = jest.fn() as any;
  });

  beforeEach(() => {
    (KubernetesApi as any).create();
  });

  afterEach(() => {
    clearGlobalMocks();
  });

  it('should exit when the name is too long', async () => {
    await previewDeployCommand.handler({
      name: '1234567890123456789012345678901234567890123456789012345678901234567890',
    } as any);
    expect(process.exit).toHaveBeenCalledWith(1);
    expect((Logger as any).instance.error.mock.calls[0][0]).toContain('Name of the namespace');
  });

  it('should use the default kube config when no flag is set', async () => {
    defaultFiles();
    const spy = jest.spyOn(KubernetesApi, 'fromDefault');
    await previewDeployCommand.handler({
      name: 'foobar',
      prefix: 'prev-',
      sourceFolder: '/',
    } as any);

    try {
      expect(spy).toHaveBeenCalled();
    } finally {
      spy.mockRestore();
    }
  });

  it('should use the given kube config content when the flag is set', async () => {
    defaultFiles();
    const spy = jest.spyOn(KubernetesApi, 'fromString');
    await previewDeployCommand.handler({
      name: 'foobar',
      prefix: 'prev-',
      sourceFolder: '/',
      kubeConfig: 'yeeeehaaaw',
    } as any);

    try {
      expect(spy).toHaveBeenCalled();
    } finally {
      spy.mockRestore();
    }
  });

  it('should create a namespace with the given name', async () => {
    defaultFiles();
    (KubernetesApi as any).instance.core.readNamespace.mockRejectedValueOnce({
      response: {
        statusCode: 404,
      },
    });

    await previewDeployCommand.handler({
      name: 'foobar',
      prefix: 'prev-',
      sourceFolder: '/',
    } as any);

    expect((KubernetesApi as any).instance.core.createNamespace.mock.calls[0][0]).toMatchSnapshot();
  });

  it('should not create the namespace when it already exists', async () => {
    defaultFiles();
    await previewDeployCommand.handler({
      name: 'foobar',
      prefix: 'prev-',
      sourceFolder: '/',
    } as any);

    expect((KubernetesApi as any).instance.core.createNamespace).not.toHaveBeenCalled();
  });

  it('should execute kubectl with the found yamls', async () => {
    defaultFiles();

    await previewDeployCommand.handler({
      name: 'foobar',
      prefix: 'prev-',
      sourceFolder: '/',
    } as any);

    expect(exec).toHaveBeenCalled();
  });
});
