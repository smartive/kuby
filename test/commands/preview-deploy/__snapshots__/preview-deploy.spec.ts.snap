// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands / preview-deploy should create a namespace with the given name 1`] = `
Object {
  "apiVersion": "v1",
  "kind": "Namespace",
  "metadata": Object {
    "name": "prev-foobar",
  },
}
`;
